Code Freeze CLI: Add Milestone command

psealock opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Adds a CLI command for adding a milestone. Since this is to be run at time of code freeze, a next release is determined and the resulting milestone will be two releases ahead of the current release.

* It gathers the next milestone from repo releases if not explicitly supplied
* It does not error out if the milestone already exists
* Has `dryRun` feature
* Sets Github outputs for use in Actions

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prerequisites

1. You must have a `GITHUB_TOKEN` environment variable available in your PATH. This needs to be set to a [personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token) with repo permissions enabled.

1. Install dependencies, `pnpm install --filter monorepo-utils`
3. Fork this repository and make sure you have a release with a valid semver tag. 
4. Try a dry run by supplying a `-d` option and point to your repo using the `--owner` and `--name` options
```
pnpm run utils code-freeze milestone --owner <owner> --name <name> -dryRun
```
5. If you're satisfied with the version number, create a milestone for real this time. Go to `https://github.com/<owner>/woocommerce/milestones` to confirm creation.
```
pnpm run utils code-freeze milestone --owner <owner> --name <name>
```
6. Run the command again to attempt creating the same milestone. It should succeed with an appropriate message.
```
pnpm run utils code-freeze milestone --owner <owner> --name <name>
```
7. Now pass in a milestone of your choosing and confirm its existance.
```
pnpm run utils code-freeze milestone --owner <owner> --name <name> --milestone "99.99.99"
```

<!-- End testing instructions -->

psealock committed:
Add milestone computation cli

samueljseay commented on tools/monorepo-utils/src/code-freeze/commands/milestone/index.ts:
```diff
@@ -0,0 +1,152 @@
+ * External dependencies
+ */
+import { Command } from '@commander-js/extra-typings';
+import chalk from 'chalk';
```
I started working on another command the other day, and I have a question, this project has not been setup to share code from cli-core like we originally talked about.

I have some suggested changes for this package:

1. Use ts-node like we do in other CLI packages to run commander commands.
2. Use the nodejs 16 TS config used in cli-core
3. Add cli-core as a dependency to the package json: `"cli-core": "workspace:*"`
4. Once you've done these steps you can use the `Logger` from `cli-core` and you should add methods to that object if you need extra functionality.

samueljseay commented on tools/monorepo-utils/src/github/__tests__/index.ts:
```diff
@@ -0,0 +1,50 @@
+} );
+
+it( 'should return the latest release version', async () => {
+	expect(
```
Nice, great to see a straightforward and useful test here.

samueljseay commented on tools/monorepo-utils/src/github/api.ts:
```diff
@@ -0,0 +1,15 @@
+import { graphql } from '@octokit/graphql';
+import { Octokit } from 'octokit';
+
+export const graphqlWithAuth = graphql.defaults( {
```
This irks me being here. cli-core defines a lot of reusable code like this, this code should go in that package? I think we should keep all the utility stuff like this contained to that package so we can all easily reuse it.

samueljseay commented on tools/monorepo-utils/src/github/api.ts:
```diff
@@ -0,0 +1,15 @@
+
+export const graphqlWithAuth = graphql.defaults( {
+	headers: {
+		authorization: `Bearer ${ process.env.GITHUB_TOKEN }`,
```
If this was to be reusable it would be better not to rely on specific process variable being present, better to pass the arg to a function and then any project can decide what env variable they use.

samueljseay commented on tools/monorepo-utils/src/github/api.ts:
```diff
@@ -0,0 +1,15 @@
+	},
+} );
+
+export const octokitWithAuth = new Octokit( {
```
If this is to be reusable, like shared amongst projects, this should allow passing the token to a function to get the instance, as in the other situation I think it's better for this kind of shared code not to make assumptions that a specific variable is present.

samueljseay commented on tools/monorepo-utils/src/github/repo.ts:
```diff
@@ -0,0 +1,42 @@
+ */
+import { graphqlWithAuth } from './api';
+
+export const getLatestReleaseVersion = async ( options: {
```
Again just wondering about why this code isn't in cli-core. Obviously we have to think about where all shared utility code should live.

samueljseay commented on tools/monorepo-utils/src/github/repo.ts:
```diff
@@ -0,0 +1,42 @@
+		.map( ( node ) => node.tagName )
+		.filter( ( tagName ) =>
+			// Remove any non-numeric and decimal tags, ie prerelease, nightly, and beta tester tags.
+			/^[\d.]+$/.test( tagName )
```
I'm starting to think this approach doesn't make sense.

Can't you just find the first release that has the `isLatest` field?

I think otherwise if we release something for another package like we do with beta tester and it has a higher version number it'll be returned instead. Your test case in jest looks like maybe the only reason it doesn't return wc-beta-tester is that 99.99.00 is not a valid version number?

psealock replied:
>Can't you just find the first release that has the isLatest field?

Thanks for the suggestion @samueljseay. Looking at the data on `woocommerce/woocommerce` with the `isLatest` field supplied appears that it will work well.

<img width="417" alt="image" src="https://user-images.githubusercontent.com/1922453/234130220-598336c8-d284-4622-8e7a-5df5bc62d4a6.png">

However, what if the Beta Tester was the last release made from the monorepo? Then `isLatest` would incorrectly surface the wrong release.

>I think otherwise if we release something for another package like we do with beta tester and it has a higher version number it'll be returned instead. Your test case in jest looks like maybe the only reason it doesn't return wc-beta-tester is that 99.99.00 is not a valid version number?

In this case the regex you commented on, `/^[\d.]+$/.test( tagName )`, tests for the presence of only digits, so releases like `nightly`, `wc-beta-tester-99.99.00` or `wc-beta-tester-99.99.0` will be filtered out and only WooCommerce Core releases will be compared.

I fixed the test sample data to have a valid version for the Beta Tester in https://github.com/woocommerce/woocommerce/pull/37723/commits/d4e8dcdddb3fb71c41ce53b4a75355e8a8ea9f30

psealock replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/37723/commits/9571f0b126afdf101f10f932c24c513697baec59.

Beta Tester and prereleases never get the `isLatest: true` property

samueljseay reviewed (changes requested):
1. I think we should discuss where shared code should go. When we introduced monorepo-utils in a recent PR I didn't notice that it wasn't using `Logger` and that it wasn't setup to consume from `cli-core` like was the original intent.

2. I think maybe `isLatest` should be a much simpler way to find the latest release.

psealock commented:
Thanks for the review @samueljseay 

I addressed your comment about simplifying `getLatestReleaseVersion`.

As we spoke about in Slack, functions like that belong in cli-core along with others you mentioned in your review. Here is an issue to track that work separately. I'll take your suggestions about using `Logger` as well as the authentication api calls.

https://github.com/woocommerce/woocommerce/issues/37947

Could we merge this as is and continue with the refactor?

samueljseay reviewed (dismissed):
LGTM, thanks for implementing those changes. I'm happy for the cli core code to handled later.

🚀

samueljseay reviewed (approved):
🚀
